[docs] Add "Open with Colab" button to documentation

guberti opened the pull request:
This pull request adds an "Open in Colab" button to **all** of TVM's tutorials. With one click, users can play around with and modify the TVM docs without having to install **anything** locally. The docs will now have this button at the top:

![image](https://user-images.githubusercontent.com/3069006/209590230-5f6500ff-0e94-4349-a3d7-5ddf8aa7bc3c.png)

----

There seems to be an issue with @tvm-bot building docs right now (see other PRs), so I don't have a live demo. You can try it locally by cloning the repo and following the steps in [docs/README.md](https://github.com/apache/tvm/tree/main/docs/README.md). You can also check out a few examples here:

- Basic functionality: [How To / Compile PyTorch models](https://colab.research.google.com/github/guberti/tvm-site/blob/asf-site/docs/_downloads/1f4943aed1aa607b2775c18b1d71db10/from_pytorch.ipynb)
- CUDA GPU support: [How To / Optimize Convolution on GPU](https://colab.research.google.com/github/guberti/tvm-site/blob/asf-site/docs/_downloads/854257a66df713b1f3f82eb3577f95e3/opt_conv_cuda.ipynb)
- microTVM compilation support [How To / Use microTVM Host-driven AOT](https://colab.research.google.com/github/guberti/tvm-site/blob/asf-site/docs/_downloads/c00933f3fbcf90c4f584d54607b33805/micro_aot.ipynb#scrollTo=n4vOD2ZJYtML)

This pull request was co-authored by @guberti, @mehrdadh, @areusch, @mkatanbaf, and @alanmacd as a hackathon project.

echuraev commented on gallery/how_to/work_with_microtvm/install_zephyr.py:
```diff
@@ -0,0 +1,29 @@
+%%shell
```
Probably it would be better to format this script as a documentation file in sphinx format?

guberti replied:
These are now reStructuredText files.

echuraev commented on apps/microtvm/poetry.lock:
```diff
@@ -1,28 +1,9 @@
 [[package]]
 name = "absl-py"
-version = "1.2.0"
+version = "1.3.0"
```
How these changes in `apps/microtvm` are related to adding "Open with Colab" button?

guberti replied:
cc @mehrdadh.

mehrdadh replied:
This was a hack, please revert it.

guberti replied:
Fixed.

echuraev commented on gallery/how_to/work_with_microtvm/install_zephyr.rst:
```diff
@@ -0,0 +1,34 @@
+Install the Prerequisites
+----------------------------
+
+    .. code-block:: bash
```
The same question here. Probably we should design this file as a documentation and add more text information? Because it looks more like a bash script for now.

guberti replied:
There is now a comment clarifying this.

echuraev commented on gallery/how_to/work_with_microtvm/micro_aot.py:
```diff
@@ -30,6 +30,29 @@
 
+######################################################################
+#
+#     .. include:: ../../../../gallery/how_to/work_with_microtvm/install_zephyr.rst
```
Tried to find where else you have included this `install_zephyr.rst`. But didn't find it. Probably you can just move the context of `install_zephyr.rst` to this file?

guberti replied:
I'd be in favor of this, but I'd love to get thoughts from @mehrdadh.

mehrdadh replied:
This file should be added to all microTVM Zephyr tutorials. I think some of them are missing in this PR. That's why we refactored it to a separate file

echuraev commented on docs/conf.py:
```diff
@@ -84,6 +84,100 @@ def git_describe_version(original_version):
+        This tutorial can be used interactively with Google Colab! You can also click
+        :ref:`here <sphx_glr_download_{1}>` to run the Jupyter notebook locally.
+
+        .. image:: https://raw.githubusercontent.com/guberti/web-data/main/images/utilities/colab_button.svg
```
While I had reviewed your PR, I found that the same image file was already used in `gallery/how_to/work_with_microtvm/micro_train.py`. So probably, you can reuse this link: https://raw.githubusercontent.com/tlc-pack/web-data/main/images/utilities/colab_button.png

guberti replied:
During the Hackathon, I redid this image to be a cleaner SVG. I've got a PR out to update this and remove the `png` version - see https://github.com/tlc-pack/web-data/pull/18.

echuraev commented on gallery/how_to/work_with_microtvm/micro_train.py:
```diff
@@ -71,7 +60,7 @@
 #     .. code-block:: bash
 #
-#       %%bash
+#       %%shell
```
What is the difference between `%%bash` and `%%shell`?

alanmacd replied:
This is some good info, I wonder if we could capture it somewhere in a comment? (not sure where would be best, so maybe not)

guberti replied:
I've added a section about this to `docs/README.md`.

echuraev commented on tests/lint/check_request_hook.py:
```diff
@@ -142,7 +142,7 @@ def find_code_block_line(lines: List[str]) -> Optional[int]:
                     actual, expected = line_info
                     print(f"{file} (misplaced hook at {actual}, expected at {expected})")
-            exit(1)
+            #exit(1)
```
Why did you comment this line?

echuraev commented on gallery/how_to/compile_models/from_keras.py:
```diff
@@ -102,8 +102,8 @@
-target = "cuda"
-dev = tvm.cuda(0)
+target = "llvm"
+dev = tvm.cpu(0)
```
I know that colab provides several runtimes (`CPU` or `GPU`), which can be selected manually. Can we automatically select preferable runtime for some scripts, e.g. where we want to use `GPU`?

echuraev reviewed (commented):
Thank you for very good improvements of TVM documentation. Looks awesome! Some comments:

guberti commented:
@echuraev thanks for the very fast review! We weren't quite done with the PR at the time, so a few of your comments are now outdated. I've addressed the others, though - thanks for the feedback!

You also asked about CPU versus GPU runtimes for Google Colab. Unfortunately, these can only be selected from the drop-down menu. However, I modified `docs/conf.py` to look for usage of `cuda` - now, we install a CUDA-enabled TVM version if and only if the tutorial requires it.

areusch commented on apps/microtvm/pyproject.toml:
```diff
@@ -98,22 +95,23 @@ onnxoptimizer = { version = "==0.2.6", optional = true }
 # Pytorch (also used by ONNX)
-torch = { version = "==1.11.0", optional = true }
-torchvision = { version = "==0.12.0", optional = true }
+torch = { version = "==1.11.0" }
```
how come you removed `optional = true` here? we use this so that you don't have to install these packages, but may opt to do so with e.g. `poetry install -E importer-torch`

guberti replied:
cc @mehrdadh who made this change

mehrdadh replied:
this was a temporary hack to be able to export the dependencies in a requirement.txt file. We should revet this change and figure out how to generate dependencies file with extra dependencies.

mehrdadh replied:
Looks like we no longer use requirement.txt file, so you can just revert changes in this file

areusch commented on docs/conf.py:
```diff
@@ -84,6 +86,148 @@ def git_describe_version(original_version):
+.. DO NOT EDIT. THIS FILE WAS AUTOMATICALLY GENERATED BY
+.. TVM'S MONKEY-PATCHED VERSION OF SPHINX-GALLERY. TO MAKE
+.. CHANGES, EDIT THE SOURCE PYTHON FILE:
+.. "{0}"
```
suggest using a named reference e.g. {SOURCE_FILE} instead of {0} 
since this comment string is so long. here and below

areusch commented on docs/conf.py:
```diff
@@ -84,6 +86,148 @@ def git_describe_version(original_version):
+
+# Google Colab allows opening .ipynb files on GitHub by appending the GitHub path to its own url.
+COLAB_URL_BASE = "https://colab.research.google.com/github"
+IPYTHON_GITHUB_BASE = "apache/tvm-site/blob/asf-site/docs/_downloads/"
```
can you comment this one?

areusch commented on docs/conf.py:
```diff
@@ -84,6 +86,148 @@ def git_describe_version(original_version):
 release = version
 
+
+def monkey_patch(module_name, func_name):
```
you might consider using `unittest.mock.patch` here since it's in the stdlib

guberti replied:
Using stdlib functions is always good, but `unittest.mock.patch` does not quite fit this use case. I can't use `@patch` as a decorator, as the decorator must be used on the function **calling** the function we wish to monkey-patch (e.g. `save_rst_example`).

I could just call `patch` as a function afterwards, but `patch` also does not include functionality for wrapping functions. It would be possible to make `patch()` work, but I think writing our own decorator is cleaner.

I've clarified the comment to better explain this.

areusch replied:
ahh i see. potentially you could theoretically override sphinx-gallery's `reset_modules` and use the [optionally-passed `when` parameter](https://github.com/sphinx-gallery/sphinx-gallery/blob/4484d35bb556bfe081156f210904abce8732ffbf/sphinx_gallery/scrapers.py#L592) to decide whether to monkey-patch or un-monkey-patch. i agree it's sort of equivalent both ways, i just wish there was a way to avoid "poisoning" sphinx-gallery permanently by importing conf.py.

guberti replied:
I agree that `reset_modules` is a valid approach, but I think going that route would make the changes less readable. Permanently altering `sphinx-gallery` is not desirable, but it may be the cleanest option other than forking `sphinx-gallery`.

areusch commented on docs/conf.py:
```diff
@@ -84,6 +86,148 @@ def git_describe_version(original_version):
+    return real_func(text, gallery_conf, target_dir, heading_levels)
+
+
+INSTALL_TVM_DEV = f"""%%shell
```
one thing i kind of like to do is 
```
"""\
%%shell
```
then it's a bit harder to forget to see the "%%shell" part

areusch commented on gallery/how_to/compile_models/from_coreml.py:
```diff
@@ -23,13 +23,12 @@
 
 .. code-block:: bash
 
-    pip install -U coremltools --user
```
i'm curious, does `--user` not work with colab? It's not the worst thing to tell folks new to managing python venvs.

guberti replied:
`--user` works as you would expect in Colab. However, the other tutorials don't use that flag for their dependencies, so I changed this to be consistent. Might be out of scope - what do you think?

areusch commented on docs/conf.py:
```diff
@@ -84,6 +86,148 @@ def git_describe_version(original_version):
+# you must request a Google Colab instance with a GPU by going to Runtime ->
+# Change runtime type -> Hardware accelerator -> GPU. If you wish to build from
+# source, see see https://tvm.apache.org/docs/install/from_source.html
+pip install apache-tvm-cu113=={version} -f https://tlcpack.ai/wheels"""
```
i wonder if `version == "*"` if this could work for CUDA_DEV too

guberti replied:
That would run fine, but I don't think that's as clear to read. I'd prefer to keep explicitly specifying `tlcpack-nightly-cu113`.

If we want to reduce the amount of copied text, we could also have one string and template it.

areusch commented on docs/conf.py:
```diff
@@ -84,6 +86,148 @@ def git_describe_version(original_version):
+    would be cleaner specify this with a config flag, but sphinx-gallery does not support this.
+    """
+
+    call_cuda_re = re.compile(r"cuda\([0-9]*\)", re.IGNORECASE)
```
could you make this a constant with a comment describing the expected line it's looking for? Also, I'm wondering if we should look for something more explicit, like a tag at the top saying:
```
# Requires: cuda115
```

or something

guberti replied:
I agree this regex approach is gross. I've added a new sphinx-gallery flag to determine whether a tutorial requires CUDA, and I've set it in all relevant tutorials:
```python
# sphinx_gallery_requires_cuda = True
```

Unfortunately, monkey-patching the extra flag is a bit of a hack, but I still think it is nicer than the regex approach.

areusch commented on docs/conf.py:
```diff
@@ -506,6 +650,12 @@ def process_docstring(app, what, name, obj, options, lines):
 from legacy_redirect import build_legacy_redirect
 
 
+def strip_ipython_magic(app, docname, source):
```
add a docstring comment just to clarify the function. should this be removing the whole line?

guberti replied:
Yep - IPython magic is not always one word (e.g. `%%writefile output.txt`) but it is always exactly one line.

areusch commented on gallery/how_to/work_with_microtvm/install_cmsis.rst:
```diff
@@ -0,0 +1,30 @@
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+Install CMSIS-NN
```
should we add some clarifying text that this is for CoLab only, in case folks run across it as a standalone doc? or is it not linked from the tutorial index and won't be found through auto-indexing?

guberti replied:
This is not linked in the tutorial index and won't be found unless someone goes digging through the code. I'll add a comment though to clarify this.

areusch commented on gallery/how_to/work_with_microtvm/micro_aot.py:
```diff
@@ -139,6 +162,12 @@
         "board": BOARD,
         "serial_number": SERIAL,
         "config_main_stack_size": 4096,
+        "cmsis_path": "/content/cmsis"
```
add parens around the ternary, here and below

guberti replied:
Replaced this to use `os.getenv` to be cleaner.

areusch commented on gallery/how_to/work_with_microtvm/micro_tflite.py:
```diff
@@ -26,109 +26,36 @@
 testing.utils.install_request_hook(depth=3)
 # sphinx_gallery_end_ignore
 
+# You can skip the following two sections (installing Zephyr and CMSIS-NN) if the following flag is False.
```
could you explain why maybe?

mehrdadh replied:
This is because it will run on CRT, so Zephyr and CMSIS-NN dependencies are not required

guberti replied:
Rewrote this comment to be clearer.

areusch commented on gallery/how_to/work_with_microtvm/micro_tflite.py:
```diff
@@ -292,7 +217,18 @@
+        "board": BOARD,
+        "serial_number": SERIAL,
+        "config_main_stack_size": 4096,
+        "cmsis_path": "/content/cmsis"
```
add parens around the ternary, here and below

areusch reviewed (commented):
thanks @guberti ! added some feedback here, mostly just formatting and clarifications

mehrdadh commented on gallery/how_to/work_with_microtvm/install_dependencies.rst:
```diff
@@ -0,0 +1,33 @@
+
+    .. code-block:: bash
+
+        %%shell
```
@guberti are these the only dependencies for microTVM tutorials?

guberti replied:
Yes - these are the only additional dependencies we need.

alanmacd commented on gallery/how_to/compile_models/from_onnx.py:
```diff
@@ -21,13 +21,12 @@
-For us to begin with, ONNX package must be installed.
-
-A quick solution is to install protobuf compiler, and
+For us to begin with, ONNX package must be installed:
```
nit: match the wording update of the other files:

```suggestion
To begin, install the ONNX package:
```
or something like that..

alanmacd commented on gallery/how_to/work_with_microtvm/micro_aot.py:
```diff
@@ -30,16 +30,40 @@
+# Installing Zephyr takes ~20 min.
+import os
+
+use_physical_hw = bool(os.getenv("TVM_MICRO_USE_HW"))
```
I think this could be worded / setup clearer for the user, since the code doesn't check `use_physical_hw` (I assume) it's unclear who is doing the skipping.

guberti replied:
Expanded on this comment a bit.

mehrdadh reviewed (commented):
sorry for delay in review, LGTM!
